text/0000-refs.md

```diff
@@ -0,0 +1,209 @@
+
+## Basic example
+
+```html
+<!-- component.html -->
+<template>
+    <div lwc:ref="foo">Hello world!</div>
+</template>
+```
+
+```js
+// component.js
+import { LightningElement } from 'lwc';
+export default class extends LightningElement {
+  renderedCallback() {
+    console.log(this.refs.foo.textContent); // 'Hello world!'
+  }
+}
+```
+
+## Motivation
+
+As of today, the only way to access DOM nodes defined in the template via JavaScript is using standard DOM APIs on `this.template` (shadow DOM) or `this` (light DOM):
+
+```js
+// shadow DOM
+this.template.querySelector('div');
+
+// light DOM
+this.querySelector('div');
+```
+
+This works, and it follows vanilla web component conventions, but it's cumbersome. It forces the developer to think of a query selector (or another technique, like `.firstChild`) to locate an element in the DOM.
+
+In the context of light DOM components, there is an additional difficulty: `this.querySelector()` may return elements from outside of the immediate template (e.g. light DOM children). So developers have to choose their selectors carefully.
+
+### Prior art
+
+The concept of "refs" exists in many other frameworks:
+
+- [React: `createRef()`](https://reactjs.org/docs/refs-and-the-dom.html)
+- [Vue: template refs](https://v3.vuejs.org/guide/composition-api-template-refs.html)
+- [Svelte: `bind:this`](https://svelte.dev/docs#template-syntax-element-directives-bind-this)
+- [Lit: `createRef()`](https://lit.dev/docs/api/directives/#ref)
+- [Angular: `elementRef`](https://angular.io/api/core/ElementRef)
+- [Stencil: `ref`](https://stenciljs.com/docs/templating-jsx#getting-a-reference-to-a-dom-element)
+
+By building on these conventions, we offer a simple and ergonomic way to solve the problem of locating elements at runtime.
+
+## Detailed design
+
+The basic design of refs is the use of the `lwc:ref` directive in templates:
+
+```html
+<template>
+    <div lwc:ref="foo"></div>
+</template>
+```
+
+...which can be accessed at runtime in a component using `this.refs`:
+
+```js
+export default class extends LightningElement {
+  renderedCallback() {
+    console.log(this.refs.foo); // `<div>` DOM node
+  }
+}
+```
+
+### `lwc:ref`
+
+`lwc:ref` can only refer to a static string. For instance, this will throw a template compiler error:
+
+```html
+<template>
+    <div lwc:ref={dynamic}></div> <!-- Not allowed! -->
+</template>
+```
+
+`lwc:ref` cannot be applied to `<template>`s, or to `<slot>`s in light DOM:
+
+```html
+<template>
+    <template if:true={foo} lwc:ref="foo"></template> <!-- Not allowed! -->
+</template>
+```
+
+```html
+<template lwc:render-mode="light">
+    <slot lwc:ref="foo"></slot> <!-- Not allowed! -->
+</template>
+```
+
+### `this.refs`
+
+`this.refs` is a plain object, with string keys and DOM node values. If a ref does not exist, then `this.refs["unknown"]` returns `undefined`.

```

Comment:
> If a ref does not exist, then `this.refs["unknown"]` returns `undefined`.

This reads like if a ref does exist then `"unknown" in this.refs` returns `false`, and if a ref (which one?) then "unknown" in this.refs` returns `true`.  Would it be more clear to write this?
> If a ref does not exist it returns `undefined`. For example, `this.refs.unknown` returns `undefined`.

Reply:
> with string keys and DOM node values

Why DOM node values and not DOM element values? How do you attach `lwc:ref` to nodes that are not elements?

Reply:
Good points, fixed.